### PR TITLE
Enable Clang static analyzer build.

### PR DESCRIPTION
**Pull request by coryan**
The changes are pretty simple, except for the tweaks to speed up
the build.  This is a fix for #10.

### CODE REVIEW

File: bigtable/CMakeLists.txt [hunk 1]
```diff
@@ -211,6 +211,14 @@ if (FORCE_SANITIZER_ERRORS)
             PRIVATE -DBIGTABLE_CLIENT_FORCE_SANITIZER_ERRORS)
 endif (FORCE_SANITIZER_ERRORS)
 
+option(FORCE_STATIC_ANALYZER_ERRORS
+    "If set enable tests that force errors detected by the static analyzer."
```
Comment (mbrukman) [inline bigtable/CMakeLists.txt]: s/set enable/set, enable/

File: ci/build-docker.sh [hunk 2]
```diff
@@ -22,10 +22,23 @@ set -eu
 mkdir -p gccpp/build-output
 cd gccpp/build-output
 
-cmake -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" ${CMAKE_FLAGS:-} ..
+CMAKE_COMMAND="cmake"
+if [ "${SCAN_BUILD}" = "yes" ]; then
+    CMAKE_COMMAND="scan-build cmake"
```
Comment (mbrukman) [inline ci/build-docker.sh]: It's 2-space indent elsewhere in the file (below).

File: ci/build-docker.sh [hunk 3]
```diff
@@ -22,10 +22,23 @@ set -eu
...
+fi
+${CMAKE_COMMAND} -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" ${CMAKE_FLAGS:-} ..
+
+# If scan-build is enabled we need to manually compile the dependencies,
+# otherwise the static analyzer finds issues in them, and there is no way to
+# ignore them.
+if [ "${SCAN_BUILD}" = "yes" ]; then
+    make -j ${NCPU} -C bigtable googleapis
```
Comment (mbrukman) [inline ci/build-docker.sh]: It's 2-space indent elsewhere in the file (below).

File: ci/build-linux.sh [hunk 4]
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
...
      --build-arg CC="${CC}" \
      --build-arg NCPU="${NCPU:-2}" \
      --build-arg BUILD_TYPE="${BUILD_TYPE:-Release}" \
      --build-arg CHECK_STYLE="${CHECK_STYLE:-}" \
+     --build-arg SCAN_BUILD="${SCAN_BUILD:-}" \
      --build-arg GENERATE_DOCS="${GENERATE_DOCS:-}" \
      --build-arg CMAKE_FLAGS="${CMAKE_FLAGS:-}" \
      -f "ci/Dockerfile.${DISTRO}" .
+
+if [ "${SCAN_BUILD:-}" = "yes" ]; then
```
Comment (mbrukman) [inline ci/build-linux.sh]: 2-space indent here as well, please — it's now codified in our `.editorconfig` so it's official! ;-)

File: bigtable/CMakeLists.txt (hunk 1 above)
Comment (coryan) [reply bigtable/CMakeLists.txt]: Done.

File: ci/build-docker.sh (hunk 2 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-docker.sh (hunk 3 above)
Comment (coryan) [reply ci/build-docker.sh]: Done.

File: ci/build-linux.sh (hunk 4 above)
Comment (coryan) [reply ci/build-linux.sh]: Yeah, and my IDE seems to forget about it every other day... grrrr.

**Comment by coryan** [comment]
PTAL.

We should wait until the PR finishes to see how long the builds take, we might want to disable the static analysis for PRs.

**Comment by mbrukman** [comment]
SGTM on disabling static analysis for PRs, just run it on the `master` branch, no need to slow down PR review/merge process.

File: ci/build-linux.sh [hunk 5]
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
...
+    sudo docker run --rm -it --volume $PWD:/d ${IMAGE}:tip \
+	 bash -c 'if [ ! -z "$(ls -1d /tmp/scan-build-* 2>/dev/null)" ]; then cp -r /tmp/scan-build-* /d/scan-build-output; fi'
+    if [ -r scan-build-output/index.html ]; then
```
Comment (mbrukman) [inline ci/build-linux.sh]: This should be indented at the level of the previous line, no?

File: ci/build-linux.sh [hunk 6]
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
...
+        echo -n $(tput setaf 1)
```
Comment (mbrukman) [inline ci/build-linux.sh]: 2-space indent?

File: ci/build-linux.sh [hunk 7]
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
...
+	echo "scan-build detected errors.  Please read the log for details."
```
Comment (mbrukman) [inline ci/build-linux.sh]: I don't follow the indentation here. Should all the `echo` lines indent together?

File: ci/build-linux.sh [hunk 8]
```diff
@@ -22,15 +22,35 @@ if [ "${TRAVIS_OS_NAME}" != "linux" ]; then
...
+	echo "To run scan-build locally, and examine the html output please"
+	echo "install and configure Docker, then run:"
+        echo
+        echo "DISTRO=ubuntu DISTRO_VERSION=17.04 SCAN_BUILD=yes NCPU=8 TRAVIS_OS_NAME=linux CXX=clang++ CC=clang ./ci/build-linux.sh"
+        echo
+        echo "the html output will be copied into the scan-build-output"
+	echo "subdirectory."
+        exit 1
+    else
+        echo -n $(tput setaf 2)
```
Comment (mbrukman) [inline ci/build-linux.sh]: 2-space indent in the `else` clause.

File: ci/build-linux.sh (hunk 5 above)
Comment (coryan) [reply ci/build-linux.sh]: `bash` is an argument to the `sudo docker` command, I thought we were indenting those by 2?
Anyway, changed the code to make this more readable.

File: ci/build-linux.sh (hunk 7 above)
Comment (coryan) [reply ci/build-linux.sh]: This was a mess, fixed.

File: ci/build-linux.sh (hunk 8 above)
Comment (coryan) [reply ci/build-linux.sh]: Fixed.

File: ci/build-linux.sh (hunk 6 above)
Comment (coryan) [reply ci/build-linux.sh]: Fixed.

File: ci/build-linux.sh (hunk 5 above)
Comment (mbrukman) [reply ci/build-linux.sh]: Sorry, misread, you're right.

**Comment by coryan** [comment]
PTAL.  I think I addressed all your comments, but I have rebased so many times I might have missed some.  Oh, and we need to wait for it to pass the tests, of course.

**Comment by coryan** [comment]
I am going to close this PR because the build is still too slow, it just went from 37 minutes to 50+ minutes and aborted on my branch.

**Comment by mbrukman** [comment]
Per offline conversation, please resurrect the part to allow folks to manually run this from the CLI, but leave out the part that automatically runs it on Travis CI — it will still be useful for us in the future to be able to easily run this on our workstations, even if it can't run automatically on Travis CI.

**Pull request closed without merging**